pallets/identity/src/lib.rs

```diff
@@ -2123,64 +2123,57 @@ impl<T: Config> CheckAccountCallPermissions<T::AccountId> for Module<T> {
+        // In such cases, the key for which we check permissions is `who`'s DID rather than `who`.
+        // TODO(Centril): Investigate the cases.
 
-        let key_did = <KeyToIdentityIds<T>>::get(who);
+        let key_did = <KeyToIdentityIds<T>>::try_get(who).ok()?;
 
         if !<DidRecords<T>>::contains_key(&key_did) {
             // The DID record is missing.
             return None;
         }
 
-        if let Some(target_did) = Context::current_identity_or::<Self>(who).ok() {
-            let target_did_record = <DidRecords<T>>::get(&target_did);
+        // Extract `who`'s DID, or bail if none.
+        let target_did = Context::current_identity_or::<Self>(who).ok()?;
+        let target_did_record = <DidRecords<T>>::get(&target_did);
+
+        // NB: Doing this check here since `get_sk_data` moves `target_did_record`
+        let is_direct_call = target_did == key_did;
+        if is_direct_call && who == &target_did_record.primary_key {
+            // It is a direct call and `who` is the primary key.
+            return Some(AccountCallPermissionsData {
+                primary_did: target_did,
+                secondary_key: None,
+            });
+        }
 
-            // NB: Doing this check here since `get_sk_data` moves `target_did_record`
-            let is_direct_call = target_did == key_did;
-            if is_direct_call && who == &target_did_record.primary_key {
-                // It is a direct call and `who` is the primary key.
-                return Some(AccountCallPermissionsData {
-                    primary_did: target_did,
-                    secondary_key: None,
-                });
-            }
+        // DIDs with frozen secondary keys (aka frozen DIDs) are not permitted to call extrinsics.
+        if Self::is_did_frozen(&target_did) {
+            // `target_did` has its secondary keys frozen.
+            return None;
+        }
 
-            let get_sk_data = |who_sk| {
-                // DIDs with frozen secondary keys (aka frozen DIDs) are not permitted to call extrinsics.
-                if Self::is_did_frozen(&target_did) {
-                    // `target_did` has its secondary keys frozen.
-                    return None;
-                }
-                target_did_record
-                    .secondary_keys
-                    .into_iter()
-                    .find(|sk| sk.signer == who_sk)
-                    .filter(|sk| sk.has_extrinsic_permission(pallet_name, function_name))
-                    .map(|sk| AccountCallPermissionsData {
-                        primary_did: target_did,
-                        secondary_key: Some(sk),
-                    })
-            };
+        let who_sk = if is_direct_call {
+            // Direct call. We'll check the permissions of `who` in `target_did`.
+            Signatory::Account(who.clone())
+        } else {
+            // Forwarded call. We'll check the permissions of `who`'s DID in `target_did`.
```

Comment:
Wondering if we really need this, or whether `!is_direct_call` is an error condition now that we don't have forwarded calls.